keps/sig-node/4816-dra-prioritized-list/README.md

```diff
@@ -0,0 +1,1022 @@
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+"Obtainability" of certain types of scarce resources is a primary concern of
+many AI/ML users. GPUs are in high demand, particularly the latest models. This
+means that workloads that use DRA to specify a need for particular types of GPUs
+may fail to schedule. In practice, a workload that needs a GPU can be written
+such that it can discover the GPUs available to it, and work with what it is
+given. A user may have a preference for the latest model, but would like to run
+the workload even if only an older model is available.
+
+Similarly, packaged workload authors may wish to configure a workload such that
+it will work well in the widest selection of available clusters. That is, a
+distributor of shared workload definitions would like to be able to specify
+alternative types of devices with which their workload will function, without
+requiring the user to modify the manifests.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+* Allow workload authors, when specifying a `ResourceClaim`, to provide a list
+  of ways to satisfy the claim, with a preference ranking.
+* Enable schedulers to evaluate those preferences and allocate devices for the
+  claim based on them.
+* Enable cluster autoscalers to evaluate those preferences and make scaling
+  choices based on them.
+* Provide some measure of ResourceQuota controls when users utilize claims with
+  these types of requests.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Enable cross-claim consistency of request choices. For example, guaranteeing
+  that all `ResourceClaim`s associated with a given `Deployment` are satisfied
+  using the same choice from the list of possible alternatives.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal adds a new type, called `RankedDeviceRequest`, which allows the
+user to list `DeviceRequest`s, exactly one of which must be satisfied. The
+`DeviceClaim` then gets a new field listing all of these such requests that must
+be satisfied. There is no change to the existing `DeviceRequest` type.
```

Comment:
This approach is the opposite of how we envisioned to add this feature. Adding a new field means that old consumers acting on a DeviceClaim will not see that new field. Instead of detecting that they are out-dated and cannot handle the claim, they will treat it like an empty claim.

Instead, the plan was to extend `DeviceRequest` with a `OneOf` slice and leave the rest of the fields unset, thus treating it like a "one-of": either the other fields are set or the new `OneOf`, but never both. This depends on `DeviceClassName` being required and current consumers checking that it is set, which they should.

We could reuse `DeviceRequest` and define that the `OneOf` slice must be empty when the `DeviceRequest` is used inside a `OneOf` (no nesting). The `DeviceRequest.Name` is meaningful the same way as you describe it in the KEP for `RankedDeviceRequest.Name`.

Reply:
> Are you saying we can relax that in validation?

I think we can. It's the "organic" evolution of the API that we discussed previously.

> Then I think we can nest the current DeviceRequest fields all under a DeviceRequestDetail which we inline. Or just duplicate the fields.

Of just use `DeviceRequest` with the constraint that the nested instances must use the fields, not the `OneOf` slice:

```
type DeviceRequest {
    ...
    // +optional
    // +oneOf=deviceRequestType
    DeviceClassName string
    
    ...
    //
    // OneOf may only be set in the entries of DeviceClaim.Requests. It must not be set
    // in DeviceRequest instances that themselves are part of a OneOf.
    //
    // +optional
    // +oneOf=deviceRequestType
    OneOf []DeviceRequest
}
```

Reusing `DeviceRequest` might make some of the implementation simpler.

Reply:
Manual validation code will have to ensure that either `DeviceClassName` or `OneOf` are set.

Reply:
I made this change, I like it. It makes the change even smaller. The one thing I did differently is use the field name `FirstOf` instead of `OneOf`, which I think is more descriptive and also won't be confused with our other use of the term "one of".

Reply:
xref my review at https://github.com/kubernetes/enhancements/pull/4871#discussion_r1794315896 
This is sort of a (non-discriminated) union, but one option is "inlined" or "unpacked".

Things I don't like:
- the declarative validation work so far does not support this
- it's unclear how to express this if ever a 3rd option arises
- admin-access should only be at the "parent" but is present in the child
- it's recursive but not REALLY (cognitive load for everyone who consumes it)

 
> Adding a new field means that old consumers acting on a DeviceClaim will not see that new field. Instead of detecting that they are out-dated and cannot handle the claim, they will treat it like an empty claim.

https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1655660015

We had a lot of discussion, and I could not find it all, but this seems relevant.  If we are ever going to have "gizmos" then we have this same problem, right?  A back-rev scheduler might not know how to hande Gizmos and treat it like they don't exist.

Reply:
One thing that's different, is this original proposal was not a "one-of". The `RankedRequests` were *in addition* to the ordinary `Requests`. So, an older client would see a partial, valid request and wouldn't know that it doesn't have all the information. To make it a one-of, we would force people to use either regular `Requests` OR `RankedRequests` in a given claim. This would work, because a RankedRequest with one entry is equivalent to a regular Request. It's just a little awkward.

Reply:
*would work, if empty claims were not considered valid, so we could differentiate